Collection, Template, Item and Query iterable ceorcision and type checking

dandesousa opened the pull request:
Ricardo,

I'd like to discuss the following changes to the interface and see if it is worthwhile.

A colleague of mine hit an issue. It was something along the following lines:

```
n = 10
col = Collection('href')
col.items = [Item() for i in range(n)]
col.to_dict()   # throws an attribute error
```

What is going on is that he unknowingly changed the items attribute from an *Array* to a *list*, which causes it to break later on the call to to_dict(). This may be a case of buyer beware, but I thought it would be nice to either make those attributes read-only or enforce the same assignment semantics as on construction. I implemented the latter in this pull request.

As a result you can now do the above and the list will be converted to an Array. Additionally, for those attributes which are sub-elements I also check to ensure they are either None, a dictionary or the type they are intended. Any other result throws a TypeError.

Please look it over when you get a chance -- and let me know what you think.

Dan

dandesousa commented:
FYI, the one weirdness this does add is that for things like Template and Error, you can now assign them using a dictionary:

```
col = Collection('href')
col.template = {'template': ...}
```

I personally don't know how much I like that but it may be desirable to some and would still error if malformed in a consistent way to from_json or the constructor form

ricardokirkner commented:
Hi Daniel,

thanks for the PR. I certainly do like the spirit of what you're trying to solve here, but I don't really like the excessive use of getter and setters. I've started a branch with a different approach to solving this in a way I like more (https://github.com/ricardokirkner/collection-json.python/pull/13#issuecomment-76522279)

I hope you won't mind me doing this, I'm not disregarding your efforts, on the contrary, I only started a branch of my own, because I was unsure of what the solution would look like.

dandesousa commented:
Ricardo,

I see what you were trying to accomplish and I understand, I think it can be cleaner. I've made a change to redo what I've done with descriptors. I think it looks better and accomplishes what I was trying to do in a more focused way. Maybe its something you can consider or incorporate in your branch.

ricardokirkner commented on collection_json.py:
```diff
@@ -6,6 +6,40 @@
 __version__ = '0.1.0'
 
 
+class ArrayProperty(object):
```
Can you add some docstring for this class?

ricardokirkner commented on collection_json.py:
```diff
@@ -6,6 +6,40 @@
+        self.value = Array(self.cls, self.name, value)
+
+
+class TypedProperty(object):
```
Can you add some docstring for this class?

ricardokirkner commented on tests.py:
```diff
@@ -325,6 +325,47 @@ def test_str(self):
+        self.assertIsInstance(collection.links, Array)
+
+        collection.items = []
+        self.assertIsInstance(collection.items, Array)
```
I'd assert for array size here too (like you do below) or in neither place

ricardokirkner commented:
This rework looks really good. I like it a lot. Just a few nitpicks

1. Instead of just testing for type conversion for ArrayProperty instances, why don't you instad check for equality with pre-built arrays, which would be testing not just the type, but also the equivalence (length and content).

2. You're only testing ArrayProperty instances, what about testing TypedPropery instances in a similar way?

The next one I'm not really sure what's best.. I'll leave it for you to decide:

3. The descriptor protocol you're using (__get__ and __set__) is documented using slightly different signatures (https://docs.python.org/3/reference/datamodel.html#object.__get__). I try to keep the original signature when overriding inherited methods, so I'd propose we do that, but I realize it makes reading the code less obvious, which makes me wonder if changing the signature here wouldn't be counterproductive. I think that at least you could keep the same arity (ie, make type non-optional like it is in the documentation).

dandesousa commented:
Ricardo,

I've addressed the issues above, documented and updated tests. Please take a look. 

1. I took the tests from your branch and expanded them slightly. I agree the array value equivalence tests unit tests are more comprehensive. IMO they are also easier to read, as they are shorter, focused and labelled better.

2. I also added TypedProperty tests for Collection.

Regarding the descriptor protocol, I think I looked at an old example, I've updated the signature to match the docs. I don't think its any less readable, I simply looked in the wrong place.

I also realized that I had a bug as I had hastily put that together as an example. In the previous example, It was being defined as a value in the class level property and that was shared by all instance of the class, so it wasn't an instance specific property (see my last two unit tests for clearer example). I changed this so that properties used the passed name as an instance level attribute name to store the data. I assumed we could store it in "_<name>". I'm not sure what I think of this, but I am open to suggestions. I also thought of adding a safety check to ensure we don't accidentally override an existing property, but I don't think its necessary.

ricardokirkner commented on .gitignore:
```diff
@@ -5,3 +5,4 @@ collection_json.egg-info/
 dist/
 .coverage
 docs/_build/
+.DS_Store
```
please remove this entry. We don't want non-project specific files in here.

ricardokirkner commented on collection_json.py:
```diff
@@ -6,6 +6,59 @@
+        self.name = name
+        self.value_attr = "_" + name
+
+    def __get__(self, instance, owner):
```
Ok, now I'm confused... this was bugging me before but now I think it's more obvious, so I'll address this clearly. Why are you using the __get__ and __set__ methods instead of __getattr__ and __setattr__ which look like they're a better match for what you're trying to do here? I think using the latter will be more correct and result in more readable code anyway.

dandesousa replied:
You mean **in place of** descriptors right? I think that the right way is to have a getter and setter with the right behavior implemented. Descriptors gets us that without verbosity of declaring properties. Its focused on just the impacted properties. 

We can use getattr and setattr, but then we pay the penalty of the check for conversion on each attribute access. Composition vs Inheritance as a design principle. 

Why does getattr, setattr of the entire object feel more correct than a get, set on the attribute itself?

ricardokirkner replied:
I didn't understand descriptors properly before. I think I grok them now, so I'm updating the comments based on that new knowledge.

ricardokirkner commented on collection_json.py:
```diff
@@ -6,6 +6,59 @@
+
+class TypedProperty(object):
+
+    """Property which is assignable only to a specific type of instance.
```
I think I prefer "Property which casts automatically to type on assignment." (reads a bit easier to me)

ricardokirkner commented on collection_json.py:
```diff
@@ -6,6 +6,59 @@
+        """
+        self.cls = cls
+        self.name = name
+        self.value_attr = "_" + name
```
I don't quite follow this change. Why is it needed to compose the attribute name? If this is a function of the name itself, we don't even need to store it as a separate instance level attribute.

ricardokirkner replied:
I think I like the approach used here a bit better: http://nbviewer.ipython.org/urls/gist.github.com/ChrisBeaumont/5758381/raw/descriptor_writeup.ipynb

That is, to avoid the instance shared data issue, use a dict keyed by instance to store the instance specific values.

dandesousa replied:
So I was excited to implement this when I saw it, as it looks like a nice solution to the problem. The post even describes using WeakRefDictionary, addressing my immediate concerns about memory leaks via cycle. 

Unfortunately I don't think the implementation would be sound in our case. For dictionaries, you need to implement __hash__ and __eq__ and preserve the property if eq(o1) == eq(o2), then hash(o1) == hash(o2). Since we override eq and ne, we must implement a decent non-trivial hash function. 

The bigger problem is that mutability would cause the hash in this function to change, meaning the next look-up could fail. Worse yet, if users created multiple objects with the same values such they were equal, it would also cause the instances to hash to the same value and change the property of both objects -- a consequence of aggregating the objects together at the class level. I think this probably wouldn't work in all but the most trivial cases.

ricardokirkner replied:
Daniel,

what about a change like this one?
http://linkode.org/ZRBW0unHfyJbllusyXJKe3/oV8avYUzsXR8Zsl1HzoVQ
This is basically the technique in the article to cope with unhashable
descriptor owners. Not sure if this is what you meant. I think we're almost
there, just need to iron out a few details to get this landed.

On Sun, Mar 1, 2015 at 9:10 PM, Daniel DeSousa <notifications@github.com>
wrote:

> In collection_json.py
> <https://github.com/ricardokirkner/collection-json.python/pull/12#discussion_r25573849>
> :
>
> > @@ -6,6 +6,59 @@
> >  __version__ = '0.1.0'
> >
> >
> > +class ArrayProperty(object):
> > +
> > +    """Property which converts from any enumerable to a typed Array instance."""
> > +
> > +    def __init__(self, cls, name):
> > +        """Constructs typed array property
> > +
> > +        :param cls type: the type of objects expected in the array
> > +        :param name str: name of the property
> > +        """
> > +        self.cls = cls
> > +        self.name = name
> > +        self.value_attr = "_" + name
>
> So I was excited to implement this when I saw it, as it looks like a nice
> solution to the problem. The post even describes using WeakRefDictionary,
> addressing my immediate concerns about memory leaks via cycle.
>
> Unfortunately I don't think the implementation would be sound in our case.
> For dictionaries, you need to implement *hash* and *eq* and preserve the
> property if eq(o1) == eq(o2), then hash(o1) == hash(o2). Since we override
> eq and ne, we must implement a decent non-trivial hash function.
>
> The bigger problem is that mutability would cause the hash in this
> function to change, meaning the next look-up could fail. Worse yet, if
> users created multiple objects with the same values such they were equal,
> it would also cause the instances to hash to the same value and change the
> property of both objects -- a consequence of aggregating the objects
> together at the class level. I think this probably wouldn't work in all but
> the most trivial cases.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ricardokirkner/collection-json.python/pull/12/files#r25573849>
> .
>

dandesousa replied:
We are close. There is a special case here that would make this work:
```
If an object defines both __get__() and __set__(), it is considered a data descriptor. Descriptors that only define __get__() are called non-data descriptors (they are typically used for methods but other uses are possible).

Data and non-data descriptors differ in how overrides are calculated with respect to entries in an instance’s dictionary. If an instance’s dictionary has an entry with the same name as a data descriptor, the data descriptor takes precedence. If an instance’s dictionary has an entry with the same name as a non-data descriptor, the dictionary entry takes precedence.
```

Since we implement both get and set, the data descriptor will always take precedence. So we can use the same name as passed in. But let me be clear, instance.__dict__[name] is just a more explicit way of getattr, setattr, so its just an equivalent way of doing what I was doing. In fact the __dict__ approach, it would fail if we decided to use, say, slot classes for your objects (something totally reasonable as a future enhancement). 

Knowing this, I'm going to implement current form with getattr, setattr -- but throwing an attribute error if the getattr doing work, and using the passed in name as the dictionary entry.

dandesousa replied:
Actually I just realized that doing this descriptor thing prohibits use from using slots, as getattr calls through to the descriptor...as long as we are ok losing this feature (slots) or changing it later if we decide its worth while, I will proceed with the other way.

ricardokirkner commented on tests.py:
```diff
@@ -11,6 +11,8 @@
     Link,
     Query,
     Template,
+    ArrayProperty,
```
Please resort these imports. The convention I use for this project (although not your fault as I haven't added this to the documentation yet) is to sort the imports asciibetically.

ricardokirkner commented:
"I changed this so that properties used the passed name as an instance level attribute name to store the data. I assumed we could store it in "_[name]". I'm not sure what I think of this, but I am open to suggestions. I also thought of adding a safety check to ensure we don't accidentally override an existing property, but I don't think its necessary -- though I'll add it if you prefer."

I think I understand now why I was confused by the value_attr attribute. I also think most of this complexity comes from the fact that you're setting the attributes as class attributes. Do you think it would be possible to refactor this code so that they're not class-level but instance level right away? That way it wouldn't be necessary to keep this extra workaround to ensure instance level isolation.

dandesousa commented:
Fixed the ordering and gitignore.

We can instead pass the name of the actual attribute to use

```
data = ArrayProperty(Data, "data", "_data")
```

But Descriptors must be defined at the class level. Therefore we need to tell it what instance attribute to store it in some how. We can either infer it OR give it explicitly.

ricardokirkner commented on collection_json.py:
```diff
@@ -6,6 +6,59 @@
 
+class ArrayProperty(object):
+
+    """Property which converts from any enumerable to a typed Array instance."""
```
I'd change this to somtehing like  'A descriptor that converts any enumerable to a type Array instance.' to me it clear this class implements the descriptor protocol

ricardokirkner commented on collection_json.py:
```diff
@@ -6,6 +6,59 @@
+        self.value_attr = "_" + name
+
+    def __get__(self, instance, owner):
+        return getattr(instance, self.value_attr, None)
```
According to the descriptor docs (https://docs.python.org/3/reference/datamodel.html?highlight=descriptor#object.__get__) this should raise an AttributeError if the attribute value can not be computed.

dandesousa commented:
Ricardo,

Made the latest changes. The approach you suggested as an alternative to getattr, setattr. The downside being that when you assume that the dict attribute exists, you can no longer use slots classes in the future as a drop it. This portion would have to be re-written to support slots classes by using a modified label (passing the slot attribute to use into the property).

ricardokirkner commented:
Alright, we're almost there. I'm perfectly happy to exchange the descriptor for the slots at this time, as we have a clear win with them and no real need for slots. We can always handle that issue when the time comes.

Before approving (everything looks really nice on the current state of this PR), can I ask you to add 2 more tests? Can you add tests to check that for TypedProperty instances, they are properly cast to the right type when they're created from dicts?

dandesousa commented:
FYI, this is ready.

ricardokirkner commented:
Great stuff! Thanks a lot for following this through to the end.
